src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,585 @@
+            }
+        }
+        log.insert("attributes", attributes);
+
+        let mut bytes =
+            serde_json::to_vec(&log).expect("Failed to encode event as json, this is a bug!");
+        bytes.push(b'\n');
+
+        Some(EncodedEvent {
+            item: PartitionInnerBuffer::new(bytes, key.into()),
+            finalizers: log.metadata_mut().take_finalizers(),
+        })
+    }
+
+    /// Generates a unique ID compatible with DD
+    fn generate_log_id(&mut self) -> String {
+        let mut id = BytesMut::with_capacity(18);
+        // timestamp in millis - 6 bytes
+        let now = Utc::now();
+        id.put_int(now.timestamp_millis(), 6);
+        // namespace + version, both 0 - 2 bytes
+        id.put_i16(0);
+
+        // host-based number - 5 bytes
+        id.put_slice(&self.host_number);
+
+        // counter - 5 bytes
+        id.put_u8(0);
+        // one padding byte
+        // 4 bytes for the counter should be more than enough - it should be unique for 1 millisecond only
+        let seq_number = self.seq_number.fetch_add(1, Ordering::Relaxed);
+        id.put_u32(seq_number); // 4 bytes
+
+        base64::encode(id.freeze())
+    }
+}
+
+fn default_encoding() -> DatadogArchivesSinkEncoding {
+    DatadogArchivesSinkEncoding {
+        key_prefix_template: Template::try_from(KEY_TEMPLATE).expect("invalid object key format"),
+        reserved_attributes: RESERVED_ATTRIBUTES.to_vec().into_iter().collect(),
+        host_number: host_number(),
+        seq_number: AtomicU32::default(),
+    }
+}
+
+fn build_s3_request(
+    req: PartitionInnerBuffer<Vec<u8>, Bytes>,
+    bucket: String,
+    path_prefix: Option<String>,
+    options: S3Options,
+) -> Request {
+    let (inner, key) = req.into_parts();
+
+    let filename = Uuid::new_v4().to_string();
+
+    let key = String::from_utf8_lossy(&key[..]).into_owned();
+    let key = format!(
+        "{}/{}{}.{}",
+        path_prefix.unwrap_or_default(),
+        key,
+        filename,
+        "json.gz"
+    )
+    .replace("//", "/");
+
+    debug!(
+        message = "Sending events.",
+        bytes = ?inner.len(),
+        bucket = ?bucket,
+        key = ?key
+    );
+
+    Request {
+        body: inner,
+        bucket,
+        key,
+        content_encoding: Compression::gzip_default().content_encoding(),
+        options: aws_s3::S3Options {
+            acl: options.acl,
+            grant_full_control: options.grant_full_control,
+            grant_read: options.grant_read,
+            grant_read_acp: options.grant_read_acp,
+            grant_write_acp: options.grant_write_acp,
+            server_side_encryption: options.server_side_encryption,
+            ssekms_key_id: options.ssekms_key_id,
+            storage_class: options.storage_class,
+            tags: options.tags,
+            content_encoding: None,
+            content_type: None,
+        },
+    }
+}
+
+/// Returns 5 first bytes of the MAC address XOR-ed with random bytes
```

Comment:
This comment isn't very helpful for understanding the function. That is, it tells me _what_ the function does but not _why_. For what reason is the MAC used and why the first 5 bytes? Why not just 5 random bytes? Consistent with my comment [here](https://github.com/timberio/vector/pull/8929/files#r700470999) you do explain in PR comments that we're building a certain sort of ID but that knowledge isn't present in the code.

Reply:
I've simplified ID generation as we agreed - no more host-dependent parts. Also documented hard constraints we have.